### PR TITLE
Add tools functionality to vLLM 

### PR DESCRIPTION
**Pull request by ArjunBhalla98**
**What this PR does / why we need it**: Currently the tools from a user request does not propagate to the vLLM backend. This PR fixes that by adding the necessary fields to `apply_chat_template`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] PyTest test suite
- [x] Manual live testing against the huggingface backend to huggingface server

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Added tools support to vLLM
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Comment by yuzisun** [comment]
@ArjunBhalla98 Seems like the commits are messed up, could you help rebase and clean up?

File: python/huggingfaceserver/tests/test_model.py [hunk 1]
```diff
@@ -507,3 +507,57 @@ async def test_input_padding_with_pad_token_not_specified(
...
+        tools=tools,
+        tool_choice="auto",
+        chat_template="{% for message in messages %}"
+        "{{ message.content }} You have these tools: {% for tool in tools %} {{ eos_token }}"
+        "{% endfor %}{% endfor %}",
+    )
+    request = ChatCompletionRequest(params=params, context={})
+    response = await bloom_model.create_chat_completion(request)
+
+    assert response.choices[0].message.content
```
Comment (yuzisun) [inline python/huggingfaceserver/tests/test_model.py:563]: assert the content ?

Comment (ArjunBhalla98) [reply python/huggingfaceserver/tests/test_model.py:563]: yeah, I think I mentioned in the slack but the output of the text generation models I tried was just mostly a repeated unrelated string which didn't really change based on the input, so I was unable to replicate the fact that the `tool_call` tag appears when testing on an actual model like QWEN, so as a sanity check I just asserted that there was output and that the parameters propagated reasonably

**Pull request closed without merging**